python/lsst/jointcal/plot_chi2.py

```diff
@@ -0,0 +1,262 @@
+            return "Starting photometric fitting...", "Updating PhotoCalib for visit:"
+        else:
+            raise RuntimeError(f"Do not understand log section type {section}.")
+
+    def _find_chi2_increase(self, chi2Data, title, threshold=1):
+        """Return True and print a message if the raw chi2 increases
+        markedly.
+        """
+        if chi2Data is None:
+            return False
+        diff = np.diff(chi2Data.raw)
+        ratio = diff/chi2Data.raw[:-1]
+        if np.any(ratio > threshold):
+            increased = np.where(ratio > threshold)[0]
+            print(title, "has increasing chi2:")
+            for x in zip(chi2Data.raw[increased], chi2Data.raw[increased + 1],
+                         ratio[increased], diff[increased]):
+                print(f"{x[0]:.6} -> {x[1]:.6} (ratio: {x[2]:.6}, diff: {x[3]:.6})")
+            print()
+            return True
+        return False
+
+    def _extract_chi2(self, logfile, section):
+        """Return the values extracted from the chi2 statements in the logfile.
+        """
+        start, end = self._get_section_start_end(section)
+        kind = []
+        chi2 = []
+        ndof = []
+        reduced = []
+        with open(logfile) as infile:
+            # Skip over lines until we get to the section start line.
+            for line in infile:
+                if start in line:
+                    break
+
+            for line in infile:
+                # Stop parsing at the section end line.
+                if end in line:
+                    break
+                match = self.matcher.search(line)
+                if match is not None:
+                    kind.append(match.group("kind"))
+                    chi2.append(match.group("chi2"))
+                    ndof.append(match.group("ndof"))
+                    reduced.append(match.group("reduced_chi2"))
+
+        # No chi2 values were found (e.g., photometry wasn't run).
+        if len(kind) == 0:
+            return None
+
+        return Chi2Data(kind, np.array(chi2, dtype=np.float64),
+                        np.array(ndof, dtype=int), np.array(reduced, dtype=np.float64))
+
+    def _plot(self, astrometry, photometry, title):
+        """Generate plots of chi2 values.
+
+        Parameters
+        ----------
+        astrometry : `Chi2Data` or None
+            The as-read astrometry data, or None if there is none too plot.
+        photometry : `Chi2Data` or None
+            The as-read astrometry data, or None if there is none too plot.
+        title : `str`
+            Title for the whole plot.
+        """
+        palette = itertools.cycle(sns.color_palette())
+
+        self.fig.clf()
+        ax0, ax1 = self.fig.subplots(ncols=2, gridspec_kw={"wspace": 0.05})
+
+        self.fig.suptitle(title)
+        # Use a log scale if any of the chi2 values are very large.
+        if max(getattr(astrometry, "raw", [0])) > 100 or max(getattr(photometry, "raw", [0])) > 100:
```

Comment:
Won't this crash if either `astrometry` or `photometry` is `None`?

Reply:
No, that's what the `[0]` default value of `getattr` is for: there will always be at least `[0]` in each of those `max()` statements.

I did add an `if astrometry is None and photometry is None: print(warning)` for that kind of case.